feat: llm-gateway: custom anthropic providers [closes ENT-1456] (#5080)
langchain-ai/docs: src/langsmith/llm-gateway-custom-providers.mdx

---
title: Custom model providers

> description: Route requests through the LLM Gateway to an endpoint you configure yourself, either OpenAI-compatible or Anthropic-compatible, such as a self-hosted open-source model served through an inference server.

Reviewer: ```suggestion
description: Route requests through the LLM Gateway to a custom OpenAI- or Anthropic-compatible endpoint, such as a self-hosted open-source model.
```
Shortening the description because it'll get truncated once too long.

---

<Note>
**Private beta:** The LLM Gateway is in private [beta](/langsmith/release-stages). Sign up for [the waitlist](https://www.langchain.com/langsmith-llm-gateway-waitlist) to get access.
</Note>

In addition to the [built-in providers](/langsmith/llm-gateway#supported-providers), the LLM Gateway can proxy requests to **any OpenAI-compatible or Anthropic-compatible endpoint** you configure yourself, such as a self-hosted open-source model served through an inference server (vLLM, Ollama, and similar).

## How it works

A custom provider is defined by a [model configuration](/langsmith/model-configurations) that you save under **Settings → Model configurations** in the [LangSmith UI](https://smith.langchain.com). The provider you select in that configuration sets the format the gateway speaks to your upstream:

| Configuration provider | Wire format | Example endpoints |
| --- | --- | --- |
| **OpenAI Compatible Endpoint** | OpenAI | `POST /v1/chat/completions`, `POST /v1/responses` |

> | **Anthropic** | Anthropic Messages | `POST /v1/messages`, `POST /v1/messages/count_tokens` |

Reviewer 2: I don't think count_tokens is supported right now? Looks easy enough to add, though.

Author: its here https://github.com/langchain-ai/langchainplus/blob/e57d997baba6233e48af9d398ff1ac1661b30f4d/smith-go/gateway/custom_openai_provider_handler.go#L78
